lib/include/openamp/virtio_mmio.h

```diff
@@ -0,0 +1,485 @@
+ * @param[in] vqs Array of pointers to vthe virtqueues used by the device.
+ *
+ * @return N/A.
+ */
+
+void virtio_mmio_register_device(struct virtio_device *vdev, int vq_num, struct virtqueue **vqs);
+
+/**
+ * @brief Setup a virtqueue structure.
+ *
+ * @param[in] dev Pointer to device structure.
+ * @param[in] idx Index of the virtqueue.
+ * @param[in] vq Pointer to virtqueue structure.
+ * @param[in] cb Pointer to virtqueue callback. Can be NULL.
+ * @param[in] cb_arg Argument for the virtqueue callback.
+ *
+ * @return pointer to virtqueue structure.
+ */
+
+struct virtqueue *virtio_mmio_setup_virtqueue(struct virtio_device *vdev,
+					      unsigned int idx,
+					      struct virtqueue *vq,
+					      void (*cb)(void *),
+					      void *cb_arg,
+					      const char *vq_name);
+
+/**
+ * @brief Notify device of virtqueue activity.
+ * @param[in] vq Pointer to virtqueue structure.
+ * @return N/A.
+ */
+
+inline void virtqueue_notify(struct virtqueue *vq)
+{
+	return vq->vq_dev->func->notify(vq);
+}
+
+/**
+ * @brief Reset virtio device.
+ * @param[in] vdev Pointer to virtio_device structure.
+ * @return N/A.
+ */
+inline void virtio_reset_vd(struct virtio_device *vdev)
+{
+	vdev->func->reset_device(vdev);
+}
+
+/**
+ * @brief Enqueues a single buffer on a virtqueue.
+ *
+ * @param[in] vq Pointer to virtqueue structure.
+ * @param[in] cookie Pointer to be returned on dequeue. Cannot be NULL.
+ * @param[in] writable Indicates if buffer is writable by device.
+ * @param[in] buffer Pointer to buffer.
+ * @param[in] len Length of the buffer.
+ *
+ * @return 0 on success.
+ * @return -ENOSPC if queue is full.
+ */
+
+int virtqueue_enqueue_buf(struct virtqueue *vq, void *cookie, int writable, char *buffer,
+			  unsigned int len);
+
+/**
+ * @brief Enqueues a transaction on a virtqueue.
+ *
+ * @param[in] vq Pointer to virtqueue structure.
+ * @param[in] cookie Pointer to be returned on dequeue. Cannot be NULL.
+ * @param[in] iov struct iovec array containing buffers to be transferred. Cannot be NULL.
+ * @param[in] readable Indicates how many buffers the device must read.
+ * @param[in] writable Indicates how many buffers the device can write.
+ *
+ * @return 0 on success.
+ * @return -ENOSPC if queue is full.
+ * @return -EMSGSIZE if the buffers can't fit in the queue.
+ * @return -EINVAL if both readable and writable are zero (no buffers to transfer).
+ */
+
+struct iovec;
+int virtqueue_enqueue(struct virtqueue *vq, void *cookie, struct iovec *iov, size_t readable,
+		      size_t writable);
+
+/**
+ * @brief Dequeues data from a virtqueue.
+ *
+ * @param[in] vq Pointer to virtqueue structure.
+ * @param[out] len Pointer to length of transferred data. Can be NULL.
+ *
+ * @return cookie parameter to virtqueue_enqueue_buf()/virtqueue_enqueue()
+ * if a buffer is available.
+ * @return NULL if there is no buffer available.
+ */
+
+void *virtqueue_dequeue(struct virtqueue *vq, uint32_t *len);
```

Comment:
I waiting to see how these 3 functions will evolve... waiting that could you create a virtio_mmio_queue.h for them?